### PR TITLE
vscode-extensions.ms-python.python: 2018.12.1 -> 2019.6.22090

### PR DESCRIPTION
**Pull request by eadwu**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by worldofpeace** [comment]
Looks good, but isn't this broken as described in https://github.com/NixOS/nixpkgs/pull/53923?

**Comment by eadwu** [comment]
Just tested on a random python file, it is broken and requires patching for the Jedi language server. Also seem to be having problems using `autopep8`.

```
Starting Jedi Python language engine.
##########Linting Output - pylint##########

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)

Error: Traceback (most recent call last):
  File "/nix/store/1xxi7wv4ih13qa49zf9wjr9hddhrs53h-vscode-insiders-extension-ms-python-python-2019.1.0/share/vscode-insiders/extensions/ms-python.python/pythonFiles/sortImports.py", line 11, in <module>
    import isort.main
  File "/nix/store/1xxi7wv4ih13qa49zf9wjr9hddhrs53h-vscode-insiders-extension-ms-python-python-2019.1.0/share/vscode-insiders/extensions/ms-python.python/pythonFiles/lib/python/isort/main.py", line 31, in <module>
    import setuptools
ModuleNotFoundError: No module named 'setuptools'


Formatting with autopep8 failed.
You could either install the 'autopep8' formatter, turn it off or use another formatter.
Error: Module 'autopep8' not installed.
```

**Comment by veprbl** [comment]
@eadwu Is this ready now?

**Comment by eadwu** [comment]
Doesn't look like it, though I'll test it again tomorrow.

**Comment by worldofpeace** [comment]
If anyone wants this to be working again, I'm pretty sure if they pick up https://github.com/NixOS/nixpkgs/pull/53923 and supply the requested changes it should work. (i think)

**Comment by eadwu** [comment]
Should've implemented the changes from that PR, will test later.

**Comment by eadwu** [comment]
Looks to be fine, at least the language server part, since tooltips are showing up. Though maybe because of my settings, I'm getting an error with `isort` which I'll get around to fixing someday.

```
Error: Traceback (most recent call last):
  File "/nix/store/a1qpngl6wcrhi6i8pc0ic9l3pnwc4bmd-vscode-insiders-extension-ms-python-python-2019.4.12954/share/vscode-insiders/extensions/ms-python.python/pythonFiles/sortImports.py", line 11, in <module>
    import isort.main
  File "/nix/store/a1qpngl6wcrhi6i8pc0ic9l3pnwc4bmd-vscode-insiders-extension-ms-python-python-2019.4.12954/share/vscode-insiders/extensions/ms-python.python/pythonFiles/lib/python/isort/main.py", line 29, in <module>
    import setuptools
ModuleNotFoundError: No module named 'setuptools'
```

File: pkgs/misc/vscode-extensions/python/default.nix
Comment (jraygauthier) [inline pkgs/misc/vscode-extensions/python/default.nix]: I believe what was meant by @worldofpeace in his review of the preceding incarnation of this PR was that it would be preferable that `languageServer` be moved in its own file (as it can be used by other packages) and taken as input.

This would indeed remove most of the complexity as the added build inputs are to my understanding meant only for the download, extraction and automated patchelf-ing of the language server:

`stdenv, fetchurl, extractNuGet, icu, curl, openssl, lttng-ust, autoPatchelfHook`

The `extractNuGet` utility could then be moved out of of the extension sub folder as well to be potentially used by others.

This is however merely a recommendation as it could indeed be done later as the need arises.

File: pkgs/misc/vscode-extensions/python/default.nix
Comment (jraygauthier) [inline pkgs/misc/vscode-extensions/python/default.nix]: You might want to add yourself as a maintainer ;)

File: pkgs/misc/vscode-extensions/python/default.nix
Comment (jraygauthier) [inline pkgs/misc/vscode-extensions/python/default.nix]: @worldofpeace : My understanding as to why `fetchNuGet` is not used is that there is no official standalone release yet as mentioned in [Add installation instructions to README · Issue #953 · microsoft/python-language-server](https://github.com/microsoft/python-language-server/issues/953#issuecomment-483067448).

 >  We do not provide official downloads for the installation yet (although download URLs are not secret but they can change). So it is recommended for clients to build their own. One of the reasons is that build we use is only tested with VS Code.

It is also **not possible** (yet) to build this be ourselves as *xbuild* is officially deprecated in favor of `msbuild` (see [Mono 5.0.0 Release Notes | Mono](https://www.mono-project.com/docs/about-mono/releases/5.0.0/#msbuild)) and does not meet the minimal requirement of this project (tried it myself).

Adding support for `msbuild` (even borrowed from debian) is an unfinished process:

 -  [mono50 doesn't include msbuild · Issue #29817 · NixOS/nixpkgs](https://github.com/NixOS/nixpkgs/issues/29817)
 -  [msbuild: init at 15.6 by leo60228 · Pull Request #43680 · NixOS/nixpkgs](https://github.com/NixOS/nixpkgs/pull/43680)
 -  [msbuild: init at 15.8 by leo60228 · Pull Request #50842 · NixOS/nixpkgs](https://github.com/NixOS/nixpkgs/pull/50842)
 
So to me, what's has been done here is fine for the time being.

**Review by jraygauthier — APPROVED** [review]
To me this would be a perfectly fine first take. 

It would be better were the language server part be moved out to its own file but even from the pt of view of its developers, this really has only been tested against vscode.

Tested the changes with a project of mine and it seems to work fine as first glance (test discovery, debugging, symbols, linting integration). 

So [Approved with Suggestions](http://neilmitchell.blogspot.com/2019/04/code-review-approve-with-suggestions.html) (**Not yet**).

UPDATE: It might be wise to wait and test this a little bit more before merging. With some more complex projects the language server seem to have trouble finding some modules when running pytest test discovery which are properly discovered when running `pytest --collect-only` from the same environment. Seems related to the isort issue above.

UPDATE2: Alright, it seems that the extensions changed quite a bit in how it works with environments. Activating the environment by **selecting the interpreter from the status bar** did the trick. For some reason, when there is a single environment, vscode does not automatically activate this environment even though your interpreter setting is explicitly set to `"python.pythonPath": "python"`. This change seems to have been persisted somewhere (most likely in one of the cache directory under `~/.config/Code`) as it did work out of the box after reloading vscode. I hope this cache won't stick with an old interpreter once I update my nix environment... Might be related to the following doc fragment:

-  [environments-and-terminal-windows](https://code.visualstudio.com/docs/python/environments#_environments-and-terminal-windows)

   >  However, launching VS Code from a shell in which a certain Python
   >  environment is activated **does not automatically activate that
   >  environment** in the default integrated terminal.

Might be of interest for other nix users:

 -  [manually-specify-an-interpreter](https://code.visualstudio.com/docs/python/environments#_manually-specify-an-interpreter)

     >  You can use an environment variable in the path setting using the
     >  syntax ${env:VARIABLE}. For example, if you've created a variable
     >  named PYTHON_INSTALL_LOC with a path to an interpreter, you can then
     >  use the following setting value:

     >  ```
     >  "python.pythonPath": "${env:PYTHON_INSTALL_LOC}",
     >  ```

     >  By using an environment variable, you can easily transfer a project
     >  between operating systems where the paths are different, just be
     >  sure to set the environment variable on the operating system first.

UPDATE 3: It would seem that this is **not totally functional yet**. When I attempt to use the "Microsoft Python Analysis Engine" instead of "Jedi" for intelliSense by setting `"python.jediEnabled": false`, I get a **"The Python Tool Server crashed 5 times in the last 3 minutes" error notification.** and nothing works anymore. @eadwu : Did you test with `"python.jediEnabled": false` ?

File: pkgs/misc/vscode-extensions/python/default.nix
Comment (worldofpeace) [reply pkgs/misc/vscode-extensions/python/default.nix]: Yep completely agree here. Perhaps a todo should be added though for when that's resolved.

**Comment by jraygauthier** [comment]
@eadwu  / @worldofpeace : I have been using the extension with `python.jediEnabled": true` (which is the default) for quite some time now and everything works fine. I personally think we should merge this in its actual state and attempt to fix the `python.jediEnabled": false` separately (a second pass).

**Comment by worldofpeace** [comment]
It would be a good idea to open issues in nixpkgs for the known regressions.

**Pull request merged**
